### PR TITLE
8-bit Adam

### PR DESCRIPTION
**Pull request by gau-nernst**
To fine-tune a pre-trained ViT-Base on [resisc45](https://huggingface.co/datasets/timm/resisc45) dataset with BF16 AMP, using default Adam optimizer from PyTorch core

```
python train.py \
  --model "timm/vit_base_patch16_224.augreg_in21k" \
  --amp bf16 \
  --optim Adam
```

To use bnb 8-bit optimizer, set `--optim AdamBnb8bit`. To use 8-bit optimizer implemented in this PR, set `--optim AdamDTQ8bit`.

Adam impl | max memory (GB) | training time | accuracy
----------|-----------------|---------------|----------
PyTorch   | 5.26            | 9m 11s        | 93.62%
bnb 8-bit | 4.78            | 9m 10s        | 93.06%
ao 8-bit  | 4.78            | 9m 15s        | 94.14%

To use wandb logging, set `--project AdamInt8` and `--run_name vit_base_bf16_amp` (change as needed).
To profile and export chrome trace, set `--profile`
To enable cosine learning rate scheduler, set `--cosine_lr_scheduler`

**Known limitation**: when learning rate is updated every step (e.g. using cosine learning rate scheduler), training speed decreases significantly. This is because we have to convert learning rate to a CUDA tensor (which incurs expensive memory transfer cost), since `torch.compile()` will treat a Python float as a constant and trigger recompile whenever the value is changed.

### CODE REVIEW

**Comment by msaroufim** [comment]
Regarding the LR scheduler limitation perhaps @mlazos knows what might be going on

**Comment by mlazos** [comment]
@gau-nernst are you compiling the optimizer? If not, the LR does not need to be a tensor. Can you share a profile? I'm not sure that copying a scalar to cuda mem should have much impact on performance E2E. Although it's possible that we are launching more kernels than necessary within the scheduler itself - this could possibly be rectified by converting to scalar before performing the calculations within the scheduler.

File: torchao/prototype/optim_8bit/adam.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                        state["max_exp_avg_sq"] = self._new_zero_buffer(p.view(-1), signed=False)
+
+                state["step"] += 1
+
+                # flatten p and grad so that torch.compile won't recompile for tensors with different ndim
+                # must explicitly convert lr to Tensor since torch.compile() will treat it as a constant
+                # if it is a python float. practically, only lr is changed during training.
+                # NOTE: if lr is change at every step, moving lr to CUDA will be a bottleneck.
+                if not isinstance(group["lr"], Tensor):
+                    group["lr"] = torch.tensor(group["lr"], device=p.device)
```
Comment (janeyx99) [inline torchao/prototype/optim_8bit/adam.py:80]: If I understand this code correctly, the code right now initiates lr as a Tensor on the device of the current param just once per group, yes?

This assumes that all params in one param group have the same device, yes?

File: train.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
+# pip install timm wandb tqdm datasets
```
Comment (msaroufim) [inline train.py:1]: let's put this in the benchmarks folder

File: torchao/prototype/optim_8bit/adam.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+class AdamDTQ8bit(Optimizer):
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        betas=(0.9, 0.999),
+        eps=1e-8,
+        weight_decay=1e-2,
+        amsgrad=False,
+        block_size=2048,
```
Comment (msaroufim) [inline torchao/prototype/optim_8bit/adam.py:17]: is this the default block size in bnb? This is quite big

File: torchao/prototype/optim_8bit/adam.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                #     group["betas"][1],
+                #     group["weight_decay"],
+                #     group["eps"],
+                # )
+
+        return loss
+
+
+# this will work with any optim state tensor subclass that implements aten.lerp.Scalar and aten.copy_.default
+@torch.compile(fullgraph=True, dynamic=True)
```
Comment (gau-nernst) [inline torchao/prototype/optim_8bit/adam.py:114]: @mlazos I'm compiling the optimizer step for each param here (called vertical fusion?). This is necessary to fuse dequant+adam_step+quant together.

@msaroufim the profile is quite big, 200MB when I ran it (or I can try compressing it). Perhaps you can run and get the profile and share with @mlazos internally? The following code will profile the first 50 iterations. Lmk if you have problems running the code
```python
python train.py \
  --model "timm/vit_base_patch16_224.augreg_in21k" \
  --amp bf16 \
  --optim AdamDTQ8bit \
  --profile
```

File: torchao/prototype/optim_8bit/adam.py (hunk 1 above)
Comment (gau-nernst) [reply torchao/prototype/optim_8bit/adam.py:80]: > If I understand this code correctly, the code right now initiates lr as a Tensor on the device of the current param just once per group, yes? And this fixes the recompile problem when lr changes value?

Yes.

> This assumes that all params in one param group have the same device, yes?

Yes. I hope this is a reasonable assumption. In what scenario this assumption would not be valid? e.g. FSDP?

File: torchao/prototype/optim_8bit/adam.py (hunk 3 above)
Comment (gau-nernst) [reply torchao/prototype/optim_8bit/adam.py:17]: Yes. It's specified in the paper (https://arxiv.org/pdf/2110.02861 section 2.1) and hard-coded in code https://github.com/TimDettmers/bitsandbytes/blob/dada530149212d64d4b69534716202659ef37ec8/csrc/kernels.cu#L4056-L4058

File: train.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            for batch in tqdm(dloader, dynamic_ncols=True, desc=f"Epoch {epoch_idx + 1}/{args.n_epochs}"):
+                if args.channels_last:
+                    batch["image"] = batch["image"].to(memory_format=torch.channels_last)
+
+                with get_amp_ctx(args.amp):
+                    loss = F.cross_entropy(model(batch["image"].cuda()), batch["label"].cuda())
+                grad_scaler.scale(loss).backward()
+
+                if args.cosine_lr_scheduler:
+                    lr = lr_schedule.get_lr(step)
```
Comment (gau-nernst) [inline train.py:177]: @mlazos @msaroufim To add context, this is the way I normally do LR schedule. Calculate LR (as a python float) and set it directly to param groups. I don't use LR scheduler from `torch.optim` because I prefer my LR schedule to be stateless (instead of stateful like in `torch.optim.lr_scheduler.LRScheduler`, which keeps track of step count internally).

LR schedulers from timm are also stateless, calculate LR as python float and should have the same problem
https://github.com/huggingface/pytorch-image-models/blob/20fe56bd9072af61d9f5404ce8b08e24ff10a807/timm/scheduler/cosine_lr.py#L81-L109
https://github.com/huggingface/pytorch-image-models/blob/20fe56bd9072af61d9f5404ce8b08e24ff10a807/timm/scheduler/scheduler.py#L77-L98

File: train.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                    for param_group in optim.param_groups:
+                        param_group["lr"] = lr
```
Comment (gau-nernst) [inline train.py:179]: I will try changing this to `param_group["lr"].copy_(lr)` instead. Maybe it helps

Comment (mlazos) [reply train.py:179]: Yes this was going to be my suggestion, where does the lr get wrapped in a tensor after getting retrieved from the scheduler? I don't see it in this code. If you were calling torch.tensor(lr, ..) this will cause an allocation of a scalar on every iteration, which is probably not good. `copy_` is a better solution since it will populate the existing memory with the current value, which should have minimal impact on the performance.

Comment (gau-nernst) [reply train.py:179]: Wrapping LR as a tensor is done on-the-fly inside the optimizer (you can scroll up to see where Jane commented).
I don't know if doing LR schedule like this is common. It's how I normally do it for my projects. We can state this as a known limitation (and the trick to solve is as you outlined here - once I confirm it helps).

Comment (gau-nernst) [reply train.py:179]: Doesn't seem to help much, though the slowdown is less than I rmb.

- Without LR schedule: 6.99 it/s
- With LR schedule using `tensor.copy_(lr)` (update every step): 6.72 it/s
- With LR schedule using python float (update every step): 6.73 it/s

File: train.py (hunk 2 above)
Comment (gau-nernst) [reply train.py]: Moved to `benchmarks/benchmark_adam_8bit.py`

File: train.py (hunk 6 above)
Comment (mlazos) [reply train.py]: Hmm interesting, yeah an E2E 4% slowdown does seem like a little much, not totally unexpected because I do expect some slowdown. Perhaps the profile will shed more light on this.

Comment (gau-nernst) [reply train.py]: [trace.json.gz](https://github.com/user-attachments/files/16075776/trace.json.gz)
I gzip-ed the file, so it's more manageable. Lmk if you prefer other format instead (for security reasons).
From my beginner's eyes, most of the optimizer's time is spent on aten::to, aten::copy and cuda stream synchronize.

**Review by msaroufim — APPROVED** [review]
🚀
